src/v3x/cqasm-values.cpp

```diff
@@ -28,35 +41,158 @@ Value promote(const Value &value, const types::Type &type) {
     if (type->as_complex()) {
-        if (const auto &const_int = value->as_const_int()) {
-            ret = tree::make<values::ConstComplex>(static_cast<ConstComplex>(const_int->value));
+        if (const auto &const_bool = value->as_const_bool()) {
+            ret = tree::make<values::ConstComplex>(static_cast<ConstComplex>(const_bool->value));
+        } else if (const auto &const_int = value->as_const_int()) {
+            ret = tree::make<values::ConstComplex>(static_cast<ConstComplex>(static_cast<double>(const_int->value)));
         } else if (const auto &const_real = value->as_const_real()) {
             ret = tree::make<values::ConstComplex>(static_cast<ConstComplex>(const_real->value));
+        } else if (const auto variable_ref = value->as_variable_ref(); variable_ref) {
+            if (types::type_check(variable_ref->variable->typ, tree::make<types::Bool>()) ||
+                types::type_check(variable_ref->variable->typ, tree::make<types::Int>()) ||
+                types::type_check(variable_ref->variable->typ, tree::make<types::Real>())) {
+                ret = value;
+            }
         }
     }
 
+    // Boolean, integer, and real arrays promote to axis
+    if (type->as_axis()) {
+        if (const auto &const_bool_array = value->as_const_bool_array()) {
+            assert(const_bool_array->value.size() == 3);
+            ret = promoteArrayValueToArrayType<ConstBoolArray, ConstRealArray, types::Real>(const_bool_array);
+        } else if(const auto &const_int_array = value->as_const_int_array()) {
+            assert(const_int_array->value.size() == 3);
+            ret = promoteArrayValueToArrayType<ConstIntArray, ConstRealArray, types::Real>(const_int_array);
+        } else if(const auto &const_real_array = value->as_const_real_array()) {
+            assert(const_real_array->value.size() == 3);
+            ret = tree::make<values::ConstRealArray>(const_real_array->value);
+        } else if (const auto variable_ref = value->as_variable_ref(); variable_ref) {
+            if (types::type_check(variable_ref->variable->typ, tree::make<types::BoolArray>()) ||
+                types::type_check(variable_ref->variable->typ, tree::make<types::IntArray>()) ||
+                types::type_check(variable_ref->variable->typ, tree::make<types::RealArray>())) {
+                assert(variable_ref->variable->typ->as_type_base());
+                assert(variable_ref->variable->typ->as_type_base()->size == 3);
+                ret = value;
+            }
+        }
+    }
+
+    // If a promotion rule was successful, copy the source location annotation from the old value to the new one
+    if (!ret.empty()) {
+        ret->copy_annotation<parser::SourceLocation>(*value);
+    }
+
     return ret;
 }
 
+/**
+ * Checks if a from_type can be promoted to a to_type.
+ */
+bool check_promote(const types::Type &from_type, const types::Type &to_type) {
+    return types::type_check(from_type, to_type) ||
```

Comment:
here, the chain of `||` can be more readable with early returns:
```suggestion
if (types::type_check(from_type, to_type)) {
  return true;
}

if (from_type->as_bool()) {
  return to_type->as_int() || to_type->as_real() || to_type->as_complex();
}

//...

return false;
```

Reply:
Yup, much better indeed. Done!